lib/AST/ReachBasedRefine.cpp

```diff
@@ -40,83 +26,116 @@ z3::expr ReachBasedRefine::GetZ3Cond(clang::IfStmt *ifstmt) {
-      std::vector<clang::Stmt *> thens({then});
-      sub->setElse(ast.CreateCompoundStmt(thens));
-      substitutions[stmt] = nullptr;
-    } else {
-      auto elif = ast.CreateIf(cond, then);
-      sub->setElse(elif);
-      sub = elif;
-      substitutions[stmt] = nullptr;
+    if (!is_unreachable) {
+      ifs.clear();
+      conds.resize(0);
+      continue;
     }
-  }
-}
 
-bool ReachBasedRefine::VisitCompoundStmt(clang::CompoundStmt *compound) {
-  // DLOG(INFO) << "VisitCompoundStmt";
-  // Create else-if cascade substitutions for IfStmts in `compound`
-  CreateIfElseStmts(GetIfStmts(compound));
-  // Apply created else-if substitutions and
-  // create a replacement for `compound`
-  if (ReplaceChildren(compound, substitutions)) {
-    std::vector<clang::Stmt *> new_body;
-    for (auto stmt : compound->body()) {
-      if (stmt) {
-        new_body.push_back(stmt);
+    conds.push_back(GetZ3Cond(if_stmt));
+
+    // Do the collected statements cover all possibilities?
+    auto is_complete{Prove(*z3_ctx, z3::mk_or(conds))};
+
+    if (ifs.size() <= 2 || !is_complete) {
+      // We need to collect more statements
+      continue;
+    }
+
+    /*
+    `body` will look like this at this point:
+
+      ...
+      i - n    : ...
+      i - n + 1: if(cond_1) { }
+      i - n + 2: if(cond_2) { }
+      ...
+      i - 1    : if(cond_n-1) { }
+      i        : if(cond_n) { }
+      ...
+
+    and we want to chain all of the statements together:
+      ...
+      i - n    : ...
+      i - n + 1: if(cond_1) { } else if(cond_2) { } ... else if(cond_n) { }
+      i - n + 2: if(cond_2) { } else if(cond_3) { } ... else if(cond_n) { }
+      ...
+      i - 1    : if(cond_n-1) { } else if(cond_n) { }
+      i        : if(cond_n) { }
+      ...
+    */
+    auto last_if{ifs[0]};
+    for (auto stmt : ifs) {
+      if (stmt == ifs.front()) {
+        continue;
+      }
+      if (stmt == ifs.back()) {
+        last_if->setElse(stmt->getThen());
+      } else {
+        last_if->setElse(stmt);
+        last_if = stmt;
       }
     }
-    substitutions[compound] = ast.CreateCompoundStmt(new_body);
+
+    /*
+    `body` will look like this at this point:
+
+      ...
+      i - n    : ...
+      i - n + 1: if(cond_1) { } else if(cond_2) { } ... else if(cond_n) { }
+      i - n + 2: if(cond_2) { } else if(cond_3) { } ... else if(cond_n) { }
+      ...
+      i - 1    : if(cond_n-1) { } else if(cond_n) { }
+      i        : if(cond_n) { }
+      ...
+
+    but since we chained all of the statements into the first, we want to remove
+    the others from the body:
+
+      ...
+      i - n    : ...
+      i - n + 1: if(cond_1) { } else if(cond_2) { } else if ...
+      ...
+    */
+    size_t start_delete{i - (ifs.size() - 2)};
+    size_t end_delete{i};
+    body.erase(body.erase(std::next(body.begin(), start_delete),
+                          std::next(body.begin(), end_delete)));
+    done_something = true;
+    break;
```

Comment:
I'd replace this `break` with adding `... && !done_something` in the loop's continuation cond. But it's only a nitpick.